packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -1530,7 +1530,16 @@ export default (superClass: Class<Parser>): Class<Parser> =>
             if (value === "global") {
               return this.tsParseAmbientExternalModuleDeclaration(nany);
             } else {
-              return this.tsParseDeclaration(nany, value, /* next */ true);
+              const oldState = this.state.clone(true);
```

Comment:
Not 100% happy with this, if there was a way to check that the next token is a line terminator it would be way better.

Reply:
We can clone state in `tsTryParseDeclare`. Since it is a try parse method, if any error is thrown in its subroutine, it should be ignored and the parser state should be restored.

Reply:
Isn't it already cloned in `tsTryParseDeclare` subroutine? Or you just mean `tsTryParse` should be used?
```js
return this.tsTryParse(() =>
  this.tsParseDeclaration(nany, value, /* next */ true),
);
```

Reply:
Although it was named `tsTryParseDeclare`, it does not restore state when error is thrown:

https://github.com/babel/babel/blob/759a82abb1f2ea64c03dc9b298aa03950690fa27/packages/babel-parser/src/plugins/typescript/index.js#L1487-L1538

> you just mean tsTryParse should be used?

That could work, just note that `tsTryParseDeclare` does not equal to `tryParse(() => tsParseDeclaration)`: The former addresses `declare ...` parsing, the latter handles general declaration, e.g.`type X = ...`.

Reply:
Definitely, but I am not sure if the entire `tsTryParseDeclare` method should be wrapped with `tsTryParse` because it is quite expensive. For the moment I will just update the current fix which is working fine.

Reply:
https://github.com/babel/babel/pull/12771#discussion_r572350118 might be useful for this.